### PR TITLE
[WIP] Lazy import the wpilib package

### PR DESCRIPTION
**Pull request by auscompgeek**
This is an attempt at lazily loading the wpilib package.

Currently only tested on CPython 3.6.

~~Still need to fix docs generation and the documentation tests…~~

### CODE REVIEW

**Comment by twinters007** [comment]
Travis will test on 3.4 and 3.5. What does the `time` command output for importing now?

File: wpilib/wpilib/__init__.py [hunk 1]
```diff
@@ -2,87 +2,44 @@
...
+        self._impl = _impl  # noqa: F821
+        self.run = run
+
+        # Re-export things used by the Python import machinery.
+        self.__name__ = __name__
+        self.__package__ = __package__
+        self.__path__ = __path__
+
+    def __getattr__(self, name):
+        if not name[0].isupper():
```
Comment (twinters007) [inline wpilib/wpilib/__init__.py]: Remind me, are drives imported with `drive.XXXDrive`?

Comment (auscompgeek) [reply wpilib/wpilib/__init__.py]: One would need to `import wpilib.drive` first.

Comment (twinters007) [reply wpilib/wpilib/__init__.py]: and wouldn't that break this?

Comment (twinters007) [reply wpilib/wpilib/__init__.py]: and wouldn't that break this?

**Comment by auscompgeek** [comment]
Well, this PR would mean that timing only an import of the wpilib package would be a near-useless metric now. A much more useful metric would be "how long does it take to go from interpreter startup to disabledInit?"

**Comment by auscompgeek** [comment]
Also Travis doesn't test on PyPy3 (where things currently still work, I believe). Might be worth adding pypy3 to the Travis config…

File: wpilib/wpilib/__init__.py (hunk 1 above)
Comment (auscompgeek) [reply wpilib/wpilib/__init__.py]: No? Why would it?

File: wpilib/wpilib/__init__.py [hunk 2]
```diff
@@ -2,87 +2,44 @@
...
+            raise AttributeError("module 'wpilib' has no attribute {!r}".format(name))
+        mod = importlib.import_module('.' + name.lower(), 'wpilib')
+        cls = getattr(mod, name)
+        setattr(self, name, cls)
+        return cls
+
+
+sys.modules['wpilib'] = WPILibLazyLoader()
```
[the remark this answers is not in the document]
Comment (auscompgeek) [reply wpilib/wpilib/__init__.py]: Uh, I guess if we put in a `__all__` it won't break IDEs. I guess that'd be useful for the docs generation too.

[the remark this answers is not in the document]
Comment (auscompgeek) [reply wpilib/wpilib/__init__.py]: Ugh, jedi doesn't listen to `__all__`.  That's annoying.  I guess this really is going to screw over IDEs.

**Comment by auscompgeek** [comment]
Uhm. I don't know why the tests are failing. Not a clue. The more I stare into the abyss, the more I think this is not a good idea.

**Comment by auscompgeek** [comment]
This experiment has been a wild ride, but I don't think this is worth it at this point. We can live for now.

**Comment by virtuald** [comment]
I'm guessing the tests are failing because networktables isn't initialized and neither is hal (presumably because imports weren't done).

File: wpilib/wpilib/__init__.py (hunk 1 above)
Comment (twinters007) [reply wpilib/wpilib/__init__.py]: cause if the first letter of the thing being imported isn't uppercase it'll throw an error?

Comment (auscompgeek) [reply wpilib/wpilib/__init__.py]: Importing a module within a package sets the attribute on the package.

**Pull request closed without merging**